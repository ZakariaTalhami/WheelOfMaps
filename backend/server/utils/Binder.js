/**
 * Binds all the methods within a class to "this"
 */
class Binder {}

/**
 * Get all the methods in the given instance
 * @param {Object} instance
 * @return {Array} - list of methods
 */
Binder.getAllMethods = function (instance) {
    var props = [];
    var obj = instance;
    while ((obj = Object.getPrototypeOf(obj)) && obj != Object.prototype) {
        props = props.concat(Object.getOwnPropertyNames(obj));
    }

    return props.sort().filter(function (e, i, arr) {
        if (e != arr[i + 1] && typeof instance[e] == "function") return true;
    });
};

/**
 * Binds the instance's methods
 * @param {Object} instance
 */
Binder.bind = function (instance) {
    Binder.getAllMethods(instance).forEach((mtd) => {
        instance[mtd] = instance[mtd].bind(instance);
    });
};

export default Binder;
