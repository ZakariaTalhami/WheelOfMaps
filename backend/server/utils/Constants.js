export const CHAPTER_IDNEX_RANGE_REGEX = /^\d{5}(-\d{5})?$/;
export const CHAPTER_RANGE_SCHEMA = {
    type: String,
    required: true,
    match: CHAPTER_IDNEX_RANGE_REGEX,
};
