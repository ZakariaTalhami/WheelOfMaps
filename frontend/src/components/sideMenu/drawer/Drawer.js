import { Flex } from "@chakra-ui/react";
import React from "react";

const DrawerWraper = (props) => (
    <Flex
        pos="absolute"
        top="0"
        left="100%"
        justify="center"
        bgColor="primaryColor"
        height="100%"
        minW="350px"
        maxW="600px"
        shadow="14px -2px 10px 4px rgb(0 0 0 / 25%)"
        transition="all 0.5s"
        transform={`translate(${props.open ? "0" : "-150%"})`}
        {...props}
    />
);

const Drawer = ({ content, onClose }) => {
    return <DrawerWraper open={!!content}>{content?.name}</DrawerWraper>;
};

export default Drawer;
