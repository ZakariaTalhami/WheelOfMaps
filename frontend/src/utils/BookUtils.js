/*
    Uility module for common functionality for Books and chapters.
*/

export const getBookChapterTitles = (book) => {
    const chapters = book.chapters || [];
    return chapters.map((chapter) => {
        const prefix =
            chapter.number === 0 ? "Prologue" : `Chapter ${chapter.number}`;

        return `${prefix}: ${chapter.title}`;
    });
};
