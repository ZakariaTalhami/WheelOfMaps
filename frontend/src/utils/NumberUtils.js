/**
 * Pads a number depending on the number of digits required
 * Returns a string with the number padded with zeros
 *
 * @param {Number} num - the desired number
 * @param {Number} places - number of digits
 * @returns {String} formated string with zero padding
 */
export function zeroPad(num, places) {
    var zero = places - num.toString().length + 1;
    return Array(+(zero > 0 && zero)).join("0") + num;
}
