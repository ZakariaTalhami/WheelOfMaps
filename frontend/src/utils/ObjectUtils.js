import _ from "lodash";

const evaluateObjectPath = (obj, path = [], defaultValue) => {
    if (_.isString(path)) path = _.split(path, ".");
    if (!_.isObject(obj) || _.isEmpty(path)) return obj || defaultValue;

    return evaluateObjectPath(obj[_.head(path)], _.drop(path), defaultValue);
};

const ObjectUtils = {
    evaluateObjectPath,
};

export default ObjectUtils;
