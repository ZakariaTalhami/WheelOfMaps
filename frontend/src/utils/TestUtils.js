import React from "react";
import { render } from "@testing-library/react";
import { ChakraProvider } from "@chakra-ui/react";
//Global Theme
import extendedTheme from "../assets/styles/extendedTheme";
// TODO: remove custom font icons, replace with react-icons
import "../assets/fonts/style.css";
// Leafletjs Styles
import "leaflet/dist/leaflet.css";
import "../assets/_leafletjs.override.scss";

const AllTheProviders = ({ children }) => {
    Object.defineProperty(window, "matchMedia", {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: false,
            media: query,
            onchange: null,
            addListener: jest.fn(), // deprecated
            removeListener: jest.fn(), // deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });

    return <ChakraProvider theme={extendedTheme}>{children}</ChakraProvider>;
};

const customRender = (ui, options) =>
    render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
