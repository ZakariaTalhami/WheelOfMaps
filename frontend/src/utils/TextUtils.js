import { Text } from "@chakra-ui/react";

/**
 * Split the summary by new line and wrap in paragraphs
 * @param {String} summary
 * @returns {Array} paragraphs of the summary
 */
export const getSummaryParagraphs = (summary) => {
    const paragraphs = summary.split("\r\n\r\n");
    return paragraphs.map((para, index) => (
        <Text
            w="100%"
            role="paragraph"
            style={{ textIndent: "1.5rem" }}
            key={index}
        >
            {para}
        </Text>
    ));
};
