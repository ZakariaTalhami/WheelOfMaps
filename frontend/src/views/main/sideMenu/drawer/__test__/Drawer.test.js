// Core
import React from "react";
// Libs
import { fireEvent, render, screen } from "../../../../../utils/TestUtils";
import { useBreakpointValue } from "@chakra-ui/react";
// Tested Component
import Drawer from "../Drawer";

jest.mock("@chakra-ui/react", () => ({
    ...jest.requireActual("@chakra-ui/react"),
    useBreakpointValue: jest.fn(),
}));

test("drawer is closed when there isnt content", () => {
    render(<Drawer content={null} />);

    expect(screen.getByTestId("side-drawer")).toHaveStyle(
        "transform: translate(-170%)"
    );
});

test("drawer is open and renders content", () => {
    const navigation = {
        Content: () => "content",
    };
    render(<Drawer content={navigation} />);

    expect(screen.getByTestId("side-drawer")).toHaveStyle(
        "transform: translate(0)"
    );
    expect(screen.getByText("content")).toBeInTheDocument();
});

test("drawer handles close event", () => {
    const onClose = jest.fn();
    const navigation = {
        Content: () => "content",
    };
    render(<Drawer content={navigation} onClose={onClose} />);

    fireEvent.click(screen.getByRole("button"));

    expect(onClose).toHaveBeenCalled();
});

describe("Mobile Drawer", () => {
    beforeEach(() => {
        useBreakpointValue.mockReturnValue(true);
    });

    test("renders in mobile view", () => {
        render(<Drawer />);

        expect(screen.getByTestId("side-drawer-mobile")).toBeDefined();
        // Drawer should be closed
        expect(screen.getByTestId("side-drawer-mobile")).toHaveStyle(
            "top: 150%"
        );
    });

    test("drawer is open", () => {
        const navigation = {
            Content: () => "content",
        };

        render(<Drawer content={navigation} />);

        expect(screen.getByTestId("side-drawer-mobile")).toBeDefined();
        expect(screen.getByTestId("side-drawer-mobile")).toHaveStyle(
            "top: 50%"
        );
    });

    test("mobile matches snapshot", () => {
        const navigation = {
            Content: () => "content",
        };
        const { container } = render(<Drawer content={navigation} />);

        expect(container).toMatchSnapshot();
    });
});

test("matches snapshot", () => {
    const navigation = {
        Content: () => "content",
    };
    const { container } = render(<Drawer content={navigation} />);

    expect(container).toMatchSnapshot();
});
