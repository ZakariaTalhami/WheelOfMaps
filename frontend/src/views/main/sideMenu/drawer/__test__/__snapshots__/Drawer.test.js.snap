// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mobile Drawer mobile matches snapshot 1`] = `
<div>
  <div
    class="css-1eimt1b"
    data-testid="side-drawer-mobile"
    open=""
  >
    <div
      class="css-1fozwq5"
    >
      <div
        class="css-5kfpbm"
      >
        <button
          class="chakra-button css-6dt22a"
          role="button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-xiiwq5"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 352 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            />
          </svg>
        </button>
      </div>
      <div
        class="css-8i4x31"
      >
        content
      </div>
    </div>
  </div>
</div>
`;

exports[`matches snapshot 1`] = `
<div>
  <div
    class="css-1fboye3"
    data-testid="side-drawer"
    open=""
  >
    <div
      class="css-1fozwq5"
    >
      <div
        class="css-5kfpbm"
      >
        <button
          class="chakra-button css-6dt22a"
          role="button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="chakra-icon css-xiiwq5"
            fill="currentColor"
            focusable="false"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 352 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            />
          </svg>
        </button>
      </div>
      <div
        class="css-8i4x31"
      >
        content
      </div>
    </div>
  </div>
</div>
`;
